utils/s2n_array.h

```diff
@@ -29,11 +34,16 @@ struct s2n_array {
 
     /* The size of each element in the array */
     uint32_t element_size;
+
+    /* The growable status of the array */
+    s2n_array_growable growable;
 };
 
 extern S2N_RESULT s2n_array_validate(const struct s2n_array *array);
 extern struct s2n_array *s2n_array_new(uint32_t element_size);
+extern struct s2n_array *s2n_array_new_with_capacity(uint32_t element_size, uint32_t capacity);
 extern S2N_RESULT s2n_array_init(struct s2n_array *array, uint32_t element_size);
```

Comment:
I think for my case it makes more sense to use array_new, since array_init would mean keeping more stuff on the connection even when the crl lookup callback isn't implemented. However, I think it makes more sense for the `new` functions to call init, and for the the initial enlarge call to happen in the inits. So I added that, and now there's both an init with capacity and a new with capacity.

Reply:
My only concern is that that's an unnecessary (and very small) allocation. If the array is on the heap, you have to allocate a few bytes for the array structure itself, then also allocate a reasonable chunk of memory for its contents. It looks like we're doing something similar for the s2n_x509_trust_store.

It would make more sense memory management wise to allocate the whole s2n_x509_validator on the heap if we're worried about its size. Then we could free the whole s2n_x509_validator after the handshake if the connection wants to optimize for memory. But by that argument, we should really move everything we free after the handshake into one structure we can allocate and free once (which I actually WOULD like to do, but that's wildly out of scope).

It also does seem like a downside of the s2n_array data structure if it requires either more memory than just a pointer + size or requires an extra heap allocation.

But it looks like we're already not being careful about s2n_x509_validator allocations, and "keeping more stuff on the connection even when the crl lookup callback isn't implemented" is an argument, so I think what you're doing is probably fine.